### PR TITLE
Fix #371 - Run emulator-based unit tests on Travis

### LINKED ISSUE #371: Emulator-based unit tests no longer work on Travis
**Summary:** 

There are issues with API Level 23-25 emulators that prevent them from running on Travis.

Here's an example of a failed build:
https://travis-ci.org/googlemaps/android-maps-utils/builds/206977302

Things go unresponsive with the recent master branch (as of https://github.com/googlemaps/android-maps-utils/commit/9be09a2e4a77a222d62d5719bcaa6413dcf373f6).  I've tried a number of resolutions in these branches/PRs (for this and another project):
* https://github.com/barbeau/android-maps-utils/tree/test-travis
* https://github.com/OneBusAway/onebusaway-android/pull/721

...but so far no luck.

It seems to boil down to these two problems with the emulators/SDK:
* API Level 23 - https://code.google.com/p/android/issues/detail?id=228113
* API Level 24/25 - https://code.google.com/p/android/issues/detail?id=230083

@stephenmcd or @broady Any chance you can try to get the Android emulator/SDK team to take a look at the above?  I'm not very familiar with how this is set up, but it may be a simple issue with the metadata associated with the SDK when fetched by something like Travis.

For now, as a workaround in PR https://github.com/googlemaps/android-maps-utils/pull/370, I've changed it to simply build the project, without spinning up an emulator and running the unit tests.  This is obviously a huge inconvenience, as you still need to run the unit tests locally to verify that they pass.

**Steps to reproduce:** 

Run a Travis build as of https://github.com/googlemaps/android-maps-utils/commit/9be09a2e4a77a222d62d5719bcaa6413dcf373f6 in the master branch, with the `.travis.yml`  config shown in https://github.com/googlemaps/android-maps-utils/blob/8f7d4b153d7281c818853474361914d3aefc0678/.travis.yml.

**Expected behavior:** 

Emulator image should be downloaded, started, and the unit tests should run via `./gradlew clean check connectedCheck -x library:signArchives -PdisablePreDex`.

**Observed behavior:** 

The emulator times out (or there is other behavior with missing ABIs, etc. depending on the API Level) - see https://travis-ci.org/googlemaps/android-maps-utils/builds/206977302 for an example, which shows:

~~~
5.61s$ echo no | android create avd --force -n test -t "android-"$ANDROID_API_LEVEL --abi $ANDROID_ABI --tag $ANDROID_TAG
Android 7.0 is a basic Android platform.
Do you wish to create a custom hardware profile [no]Created AVD 'test' based on Android 7.0, Google apis ARM (armeabi-v7a) processor,
with the following hardware config:
hw.lcd.density=240
hw.ramSize=512
vm.heapSize=48
before_script.2
0.00s$ emulator -avd test -no-skin -no-window &
$ android-wait-for-emulator
emulator: WARNING: the -no-skin flag is obsolete. to have a non-skinned virtual device, create one through the AVD manager
Creating filesystem with parameters:
    Size: 69206016
    Block size: 4096
    Blocks per group: 32768
    Inodes per group: 4224
    Inode size: 256
    Journal blocks: 1024
    Label: 
    Blocks: 16896
    Block groups: 1
    Reserved block group size: 7
Created filesystem with 11/4224 inodes and 1302/16896 blocks
pulseaudio: pa_context_connect() failed
pulseaudio: Reason: Connection refused
pulseaudio: Failed to initialize PA contextaudio: Could not init `pa' audio driver
emulator: Listening for console connections on port: 5554
emulator: Serial number of this emulator (for ADB): emulator-5554
[140504510871296]:WARNING:./android/base/files/IniFile.cpp:158:Failed to process .ini file /home/travis/.android/emu-update-last-check.ini for reading.
Your emulator is out of date, please update by launching Android Studio:
 - Start Android Studio
 - Select menu "Tools > Android > SDK Manager"
 - Click "SDK Tools" tab
 - Check "Android SDK Tools" checkbox
 - Click "OK"
[140503899166464]:WARNING:./android/base/files/IniFile.cpp:158:Failed to process .ini file /home/travis/.android/emu-update-last-check.ini for reading.
Waiting for emulator to start
Waiting for emulator to start
Waiting for emulator to start
Waiting for emulator to start
Waiting for emulator to start
Waiting for emulator to start
Waiting for emulator to start
Waiting for emulator to start
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
~~~

**Device and Android version:** 

Above example is using API Level 23 emulator - see https://github.com/googlemaps/android-maps-utils/blob/8f7d4b153d7281c818853474361914d3aefc0678/.travis.yml for exact config.

**Screenshots:** 

N/A

### PR DESCRIPTION
**Pull request by barbeau**
### tl;dr

Currently, Travis CI simply builds the library but does not run any of the unit tests. This is because the unit tests are instrumented, and therefore require a running emulator or device to execute. And getting emulators to run on Travis is hard (see https://github.com/googlemaps/android-maps-utils/issues/371).

After a lot of testing, in this PR I've managed to get the magical Travis incantations for running emulators to work. 

We still have one unit test  (`GeoJsonPointStyleTest.testIcon()`) failing on Travis (but not locally) - see **TODO** section below for our options.

### Details

I've tested emulators for a number of API Levels (15-19, and 21-29).

Several emulators (15, 16, 19, 29) are broken on Travis - they refuse to start or hang during startup.

Emulators 17, 18, and 21-28 start up correctly and execute the unit tests - so that's good!

But, emulators 17, 18, and 21-25 have two failing tests on Travis:
* `GradientTest.testInterpolateColor()`
* `GeoJsonPointStyleTest.testIcon()`.

The [`GradientTest.testInterpolateColor()` failure](https://travis-ci.org/CUTR-at-USF/android-maps-utils/jobs/606091540#L4128) is:

~~~
07:32:38 V/InstrumentationResultParser: Error in testInterpolateColor(com.google.maps.android.heatmaps.GradientTest):
07:32:38 V/InstrumentationResultParser: java.lang.AssertionError: expected:<-65434> but was:<-65433>
07:32:38 V/InstrumentationResultParser: at org.junit.Assert.fail(Assert.java:88)
07:32:38 V/InstrumentationResultParser: at org.junit.Assert.failNotEquals(Assert.java:834)
07:32:38 V/InstrumentationResultParser: at org.junit.Assert.assertEquals(Assert.java:645)
07:32:38 V/InstrumentationResultParser: at org.junit.Assert.assertEquals(Assert.java:631)
07:32:38 V/InstrumentationResultParser: at com.google.maps.android.heatmaps.GradientTest.testInterpolateColor(GradientTest.java:54)
~~~

Fortunately, this issue doesn't seem to exist on emulators 26-28, and `GradientTest.testInterpolateColor()` passes on these emulators.

### TODO

So, on emulators 26-28 that leaves us with [1 test failure](https://travis-ci.org/CUTR-at-USF/android-maps-utils/builds/606209986#L4256), `GeoJsonPointStyleTest.testIcon()`:

~~~
09:33:02 V/InstrumentationResultParser: There was 1 failure:
09:33:02 V/InstrumentationResultParser: 1) testIcon(com.google.maps.android.data.geojson.GeoJsonPointStyleTest)
09:33:02 V/InstrumentationResultParser: java.lang.NullPointerException: IBitmapDescriptorFactory is not initialized
09:33:02 V/InstrumentationResultParser: at com.google.android.gms.common.internal.Preconditions.checkNotNull(Unknown Source:11)
09:33:02 V/InstrumentationResultParser: at com.google.android.gms.maps.model.BitmapDescriptorFactory.zzg(Unknown Source:2)
09:33:02 V/InstrumentationResultParser: at com.google.android.gms.maps.model.BitmapDescriptorFactory.defaultMarker(Unknown Source:22)
09:33:02 V/InstrumentationResultParser: at com.google.maps.android.data.geojson.GeoJsonPointStyleTest.testIcon(GeoJsonPointStyleTest.java:66)
~~~

This issue is related to the Google Maps API not being initialized prior to using it, and the solution is to call [`MapsInitializer.initialize()`](https://developers.google.com/android/reference/com/google/android/gms/maps/MapsInitializer) first.

However, we're already doing this in [`GeoJsonPointStyleTest.setUp()`](https://github.com/googlemaps/android-maps-utils/blob/master/library/src/androidTest/java/com/google/maps/android/data/geojson/GeoJsonPointStyleTest.java#L19):

~~~
@Before
public void setUp() {
	MapsInitializer.initialize(InstrumentationRegistry.getInstrumentation().getTargetContext());
	pointStyle = new GeoJsonPointStyle();
}
~~~

So, I'm not sure why this test is failing on Travis.

Unless someone else has ideas for how to get this test to pass on Travis, one option is to skip `GeoJsonPointStyleTest.testIcon()` *only* when running on Travis.

In the past I've tried to configure skipping tests via environmental variables on Travis, but this hasn't worked for me:

~~~
// Don't run this test on Travis CI
String runningOnTravis = System.getenv("RUNNING_ON_TRAVIS");
if (runningOnTravis != null && runningOnTravis.equals("true")) {
	Assume.assumeTrue("Skipping GeoJsonPointStyleTest.testIcon() - this is expected behavior when running on Travis CI", false);
}
~~~

`runningOnTravis` is always `null`, so the Travis build doesn't seem to be picking up the variable from the Travis web console config.

Another option is to simply comment out the test for now.

@jpoehnelt and @amuramoto do you have any preferences on how to address the failing test on CI, and any suggestions on how to configure Travis with an environmental variable that can be read on Android?

#### References

* https://travis-ci.org/ankidroid/Anki-Android/builds/516486599
* https://github.com/ankidroid/Anki-Android/blob/master/.travis.yml
* https://github.com/OneBusAway/onebusaway-android/blob/ec5a49f0698e2fe7c00e970a788d3fe94a07feb5/.travis.yml
* https://github.com/barbeau/gpstest/blob/master/.travis.yml

### CODE REVIEW

**Comment by barbeau** [comment]
Hmmm...based on https://travis-ci.org/googlemaps/android-maps-utils/builds/607306060, it appears sometimes this configuration errors with:

~~~
> Task :library:connectedDebugAndroidTest
Skipping device 'test(AVD)' for 'library:': Unknown API Level
 > : No compatible devices connected.[TestRunner] FAILED 
Found 1 connected device(s), 0 of which were compatible.
07:52:19 I/XmlResultReporter: XML test result file generated at /home/travis/build/googlemaps/android-maps-utils/library/build/outputs/androidTest-results/connected/TEST-TestRunner-library-.xml. Total tests 1, failure 1, 
> Task :library:connectedDebugAndroidTest FAILED
~~~

I'm looking into it.

**Comment by jpoehnelt** [comment]
Regarding the envar skip, try just System.getenv("TRAVIS")

**Comment by barbeau** [comment]
@jpoehnelt I tried this:

~~~
        if (System.getenv("TRAVIS") != null) {
            Assume.assumeTrue("Skipping GeoJsonPointStyleTest.testIcon() - this is expected behavior on Travis CI (#573)", false);
            return;
        }
~~~

...but the test still executes on Travis:
https://travis-ci.org/googlemaps/android-maps-utils/builds/607323180#L2862

**Comment by jpoehnelt** [comment]
I would just comment out the test and an a bug for now if it proves too difficult to ignore in travis.

**Comment by amuramoto** [comment]
😻

**Comment by barbeau** [comment]
@jpoehnelt I left them in the `.travis.yml`, but commented out for now. To my knowledge at this time there isn't anything that would differentiate a build between APIs 26-28, so I didn't enable them to avoid burning unnecessary data center time (each build runs around 13 minutes :deciduous_tree: :recycle:). If you have a strong preference for enabling them I certainly can.

**Comment by friederbluemle** [comment]
Nice work @barbeau 👏

**Comment by barbeau** [comment]
For the record, it looks like this is the reason why the `GradientTest.testInterpolateColor()` test fails on emulators 17, 18, and 21-25 - it's an implementation issue with `android.graphics.Color.RGBToHSV()`:
https://stackoverflow.com/a/28070310/937715

**Comment by barbeau** [comment]
And, the GradientTest.testInterpolateColor() issue isn't specific to Travis - I can repro it on my desktop with an API 24 emulator.
